src/js/views/Project/Configuration/EditSSMParam.jsx

```diff
@@ -0,0 +1,248 @@
+import PropTypes from 'prop-types'
+import React, { useContext, useEffect, useState } from 'react'
+import { Form } from '../../../components'
+import { compare } from 'fast-json-patch'
+import { Trans, useTranslation } from 'react-i18next'
+import { httpDelete, httpPatch, httpPost } from '../../../utils'
+import { Context } from '../../../state'
+
+function EditSSMParam({
+  param,
+  project,
+  pathPrefix,
+  onCancel,
+  onError,
+  onSuccess
+}) {
+  const [globalState] = useContext(Context)
+  const { t, i18n } = useTranslation()
+  const [saving, setSaving] = useState(false)
+  const [name, setName] = useState(param.name.replace(pathPrefix, ''))
+  const [values, setValues] = useState(
+    Object.fromEntries(
+      project.environments.map((environment) => [
+        environment,
+        {
+          value: param.values[environment]?.value,
+          type: param.values[environment]?.type
+        }
+      ])
+    )
+  )
+
+  function deleteThenPost(nameSuffix) {
+    const createParams = async (onSuccess) => {
+      const newValues = {}
+      for (const [environment, { type, value }] of Object.entries(values)) {
+        if (value) {
+          newValues[environment] = { type, value }
+        }
+      }
+      const response = await httpPost(
+        globalState.fetch,
+        new URL(
+          `/projects/${project.id}/configuration/ssm`,
+          globalState.baseURL
+        ),
+        {
+          name,
+          values: newValues
+        }
+      )
+      if (!response.success) {
+        setSaving(false)
+        onError(
+          response?.responseBody?.detail
+            ? response.responseBody.detail
+            : response.data
+        )
+      } else {
+        onSuccess()
+      }
+    }
+
+    const deleteParams = async (onSuccess) => {
+      setSaving(true)
+      const deleteEnvironments = []
+      for (const [environment, { value }] of Object.entries(param.values)) {
+        if (value) {
+          deleteEnvironments.push(environment)
+        }
+      }
+      const response = await httpDelete(
+        globalState.fetch,
+        new URL(
+          `/projects/${project.id}/configuration/ssm`,
+          globalState.baseURL
+        ),
+        {
+          name: nameSuffix,
+          environments: deleteEnvironments
+        }
+      )
+      if (!response.success) {
+        setSaving(false)
+        onError(
+          response?.responseBody?.detail
+            ? response.responseBody.detail
+            : response.data
+        )
+      } else {
+        onSuccess()
+      }
+    }
+
+    deleteParams(() => {
+      createParams(() => onSuccess()).catch((error) => onError(error))
+    }).catch((error) => onError(error))
```

Comment:
Is there a benefit to reusing the `onSuccess` name and chaining as opposed to having `deleteParams` invoke `createParams` explicitly and `createParams` invoke the outer `onSuccess`?

```javascript
function deleteThenPost() {
  const createParams = async () => {
    //...
    if (!response.success) {
      //...
    } else {
      onSuccess()
    }
  }
  const deleteParams = async() => {
    //...
    if (!response.success) {
      //...
    } else {
      await createParams()
    }
  }
  try {
    await deleteParams()
  } catch (error) {
    onError(error)
  }
}
```

Reply:
This confused me until I figured out that `onSuccess` wasn't the outer name even though `onError` did

Reply:
No, your suggestion is more clear. The only wrinkle is you can't call `await deleteParams()` from the not-async `deleteThenPost`, but that's ok, we can just call `deleteParams().catch((error) => onError(error))` instead.